Activity 4 -- Tim Nikiforov

iefomit opened the pull request:
<!--
PLEASE CHECK THE PR GUIDLINES:
 https://docs.google.com/document/d/14ZIhQRlUPK0wl5_AmM80N2D2KJUd6k_7I6isLS89YKU/edit
These are the things we are going to be checking your PR has, if it doesn't your PR will be rejected!
-->

## Overview
<!-- A general very short overview of the PR -->
This PR will map buttons to make robot do specific tasks


## Activity Finished
<!-- The activity you finished before making this PR -->
Activity 4

## Tests Ran
<!-- Explicity list the exact tests you ran. These should be the tests listed in the Testing and Evaluation section of your design doc, and of every merged design doc. List the actual result of the test and how it was within the measurement specified by the test. -->
Simulated robot, worked


## Additional Notes
<!-- Anything else that will help the reviewer understand your PR -->

iefomit committed:
Initialize motors

iefomit committed:
Update Drivetrain.java

iefomit committed:
Merge branch 'main' into origin/Tim

iefomit committed:
Finished Activity 1

iefomit committed:
Finished 100% Activity 1

iefomit committed:
Merge branch 'main' into origin/Tim

iefomit committed:
Finished 101% Activity 1

Set the idle mode on both motors to brake

iefomit committed:
Finished 2.1

Completed Drivetrain Sim

iefomit committed:
Finishing 2.2

iefomit committed:
Finished 2.2

iefomit committed:
Finishing 2.3

Mad new subsystem

iefomit committed:
Fixing Things 2.3

iefomit committed:
Finished 2.3

iefomit committed:
Finished 2.4

iefomit committed:
Finished 101% Activity 2.4

iefomit committed:
Finished 3.1

rostovets committed:
Finished 3.2

rostovets committed:
Autonomous Square

Made an autonomous square program that moves the robot in a counter-clockwise square movement that can be launched from Shuffleboard in simulation

rostovets committed:
Finished 3.3

rostovets committed:
debug activity 2

iefomit committed:
Fixing Activity 3

iefomit committed:
Fixing more Activity 3

iefomit committed:
fixing

iefomit committed:
Still fixing (WILL change Rotate Command)

iefomit committed:
fixing more

iefomit committed:
maybe done

iefomit committed:
finishing activity 3

iefomit committed:
Finished Activity 3

iefomit committed:
100% finished activity 3

Fixed issue with subsystem motor not going to setpoint correctly in sim

iefomit committed:
gitignore

iefomit committed:
Updated tolerance check

iefomit committed:
Added Auto Button Commands 4.1

iefomit committed:
Finished 4.1

iefomit committed:
Finished 4.2

iefomit committed:
Finished 4.3

Kyle-Eldridge commented on src/main/java/frc/robot/controls/BaseDriverConfig.java:
```diff
@@ -10,7 +10,7 @@
 public abstract class BaseDriverConfig {
 
-    private final Drivetrain drive;
+    protected final Drivetrain drive;
```
Drivetrain should be private. To access it, use getDrivetrain().

Saara21 replied:
Yeah, so what protected final does is is that it's accessible within the class and its subclasses, whereas private final means the field is only accessible within the class itself. So with protected final you can have subclass access while still stopping external access from outside the class, while private final restricts both subclass and external access. We typically use private because we want to limit access strictly to the defining class, since generally there's better encapsulation—it's kept internal to the class and can’t be changed by subclasses or external code.

iefomit replied:
fixed it!

Kyle-Eldridge commented on src/main/java/frc/robot/controls/GameControllerDriverConfig.java:
```diff
@@ -1,7 +1,22 @@
+import edu.wpi.first.wpilibj2.command.PrintCommand;
+import edu.wpi.first.wpilibj2.command.RunCommand;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import edu.wpi.first.wpilibj2.command.Subsystem;
```
Remove all unused imports in all files, including this one.

iefomit replied:
done

Kyle-Eldridge commented on src/main/java/frc/robot/controls/GameControllerDriverConfig.java:
```diff
@@ -12,22 +27,68 @@
+    controller.get(Button.LB).onTrue(new AutoForwardCommand(drive, -1.0, 3.0));
+    controller.get(Button.RB).onTrue(new AutoForwardCommand(drive, 1.0, 3.0));
+    controller.get(Button.B).onTrue(new AutoSquareCommand(drive));
+    controller.get(Button.X).onTrue(new AutoTurnCommand(drive, 152.17)); //turns 180 degrees (angles for some reason broken)
```
The reason angles are broken is because the robot is still rotating when it reaches the setpoint. As you noticed while using Bang-Bang, it passes the setpoint by a lot. You don't need to fix this; this is just the reason why you need to set the lower angle. Also, how did you find that the angle should be 152.17? Did you tune it to the nearest hundredth of a degree?

iefomit replied:
yea, tuned it manually)

Kyle-Eldridge commented on src/main/java/frc/robot/controls/GameControllerDriverConfig.java:
```diff
@@ -12,22 +27,68 @@
+    // SequentialCommandGroup: Drive forward, then turn 45 degrees
+    controller.get(Button.RIGHT_JOY).onTrue(new SequentialCommandGroup(
+      new RunCommand(() -> drive.arcadeDrive(0, 0), drive),    // stop
+      new AutoForwardCommand(drive, 1.0, 3.0)                  // drive forward
```
Your comment says this will drive forward and then turn 45 degrees, but that isn't what this command does. There are a few problems with this command:
<ol>
<li>You don't need to stop the drivetrain before making it move. You can completely remove the RunCommand.</li>
<li>RunCommands never end. This is why you use them for continuous things like driving or spinning while a button is held. If you want to do something that only runs once, like stopping the drivetrain at the end of a command, use InstantCommand. To run something for a specified time, use .withTimeout().</li>
<li>You don't have any command in this group to turn the robot 45 degrees, and you need another command if you remove the RunCommand.</li>
</ol>

iefomit replied:
should work now

Kyle-Eldridge commented on src/main/java/frc/robot/controls/GameControllerDriverConfig.java:
```diff
@@ -12,22 +27,68 @@
+
+    // ConditionalCommand: Drive forward if LEFT_TRIGGER_BUTTON is pressed, otherwise turn
+new Trigger(controller.LEFT_TRIGGER_BUTTON).onTrue(
+    new ConditionalCommand(
```
If your Trigger uses LEFT_TRIGGER_BUTTON, your condition cannot also be LEFT_TRIGGER_BUTTON. You can leave this as new Trigger() or use controller.get() instead. They do the same thing for BooleanSuppliers. Also, since LEFT_TRIGGER_BUTTON is already a BooleanSupplier, which is a type of lambda expression, you do not need to include the ()-> or the .getAsBoolean().

Kyle-Eldridge replied:
I think you still have RIGHT_TRIGGER_BUTTON and a few of the DPad buttons left to use as conditions, or you could change the Trigger to use a DPad button.

iefomit replied:
should work now

Kyle-Eldridge commented on src/main/java/frc/robot/controls/GameControllerDriverConfig.java:
```diff
@@ -12,22 +27,68 @@
+);
+
+    // PrintCommand: Print message when DPAD_LEFT is pressed
+    controller.get(GameController.DPad.UP).onTrue(new PrintCommand("DPAD_UP button pressed!"));
```
Code doesn't match comment. Change one of them.

iefomit replied:
fixed it

Kyle-Eldridge commented on src/main/java/frc/robot/controls/GameControllerDriverConfig.java:
```diff
@@ -12,22 +27,68 @@
+
+    // WaitUntilCommand: Wait until RIGHT_TRIGGER_BUTTON is pressed, then turn
+    controller.get(GameController.DPad.DOWN).onTrue(new WaitUntilCommand(
+        () -> controller.RIGHT_TRIGGER_BUTTON.getAsBoolean())
```
Same comment as above. RIGHT_TRIGGER_BUTTON is already a BooleanSupplier.

iefomit replied:
changed it

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,103 @@
+        motor = new TalonFX(motorID);
+
+        armSim = new SingleJointedArmSim(
+            DCMotor.getFalcon500(1), // assuming Falcon500 motor
```
armSim should only be defined if the robot base isSimulation. You can choose whether to always define the mechanism or not.

iefomit replied:
fixed it

Kyle-Eldridge reviewed (changes requested):
You are almost done, but you still need to make a few small changes.

Saara21 commented on src/main/java/frc/robot/commands/AutoForwardCommand.java:
```diff
@@ -0,0 +1,47 @@
+    public boolean isFinished() {
+        return timer.hasElapsed(duration);
+    }
+}
```
Great job!

iefomit committed:
Fixing Activity 4

iefomit committed:
Deleted an import

Kyle-Eldridge commented on src/main/java/frc/robot/controls/GameControllerDriverConfig.java:
```diff
@@ -52,42 +49,43 @@ public void configureControls() {
+      new AutoForwardCommand(getDrivetrain(), 1.0, 2.0),
+      new AutoTurnCommand(getDrivetrain(), -90),
+      controller.get(controller.LEFT_TRIGGER_BUTTON)
     )
```
It still uses LEFT_TRIGGER_BUTTON. The condition cannot be the same as the trigger, since this would make it always true. You also don't need controller.get() here, because controller.LEFT_TRIGGER_BUTTON (or controller.RIGHT_TRIGGER_BUTTON) is already the right type. controller.get() converts it to a Trigger, and this command takes a BooleanSupplier.

iefomit replied:
just updated it

Kyle-Eldridge replied:
Now you aren't scheduling anything on the button press. The code structure should be:
controller.get(controller.LEFT_TRIGGER_BUTTON).onTrue(
new ConditionalCommand(
command1,
command2,
condition
));
condition should be a BooleanSupplier, such as controller.RIGHT_TRIGGER_BUTTON or, if something can't efficiently be converted to a BooleanSupplier, a Trigger like controller.get(DPad.something). You could also use something other than a button as a condition, like ()->getDrivetrain().getPose().getX()>5. If something is already a BooleanSupplier, like LEFT_TRIGGER_BUTTON, you don't need to use controller.get() to convert it to a Trigger.

Kyle-Eldridge replied:
You still need the controller.get(controller.LEFT_TRIGGER_BUTTON).onTrue before the command so it is scheduled when it becomes true.

iefomit replied:
Just added the scheduler. I also modified the condition you mentioned, as it lets allows the condition to actually have two cases. It's starting to make more sense now—thanks for the clarification!

iefomit committed:
Modified Conditional Command -- Activity 4

iefomit committed:
+ Scheduler, modified angles, condition, comments

Added the scheduler, modified the condition, angles, and comments to be more accurate.
